flytekit/core/context_manager.py

```diff
@@ -243,268 +194,353 @@ class Mode(Enum):
+        return self.new_builder().with_execution_state(es)
 
-    @property
-    def user_space_params(self) -> Optional[ExecutionParameters]:
-        if self._user_space_params is not None:
-            return self._user_space_params
-        elif self._parent is not None:
-            return self._parent.user_space_params
-        else:
-            raise Exception("No user_space_params initialized")
+    def with_compilation_state(self, c: CompilationState) -> Builder:
+        return self.new_builder().with_compilation_state(c)
 
-    @property
-    def execution_state(self) -> Optional[ExecutionState]:
-        return self._execution_state
+    def with_new_compilation_state(self) -> Builder:
+        return self.with_compilation_state(self.new_compilation_state())
 
-    @contextmanager
-    def new_execution_context(
-        self,
-        mode: ExecutionState.Mode,
-        additional_context: Dict[Any, Any] = None,
-        execution_params: Optional[ExecutionParameters] = None,
-        working_dir: Optional[str] = None,
-    ) -> Generator[FlyteContext, None, None]:
-        # Create a working directory for the execution to use
-        working_dir = working_dir or self.file_access.get_random_local_directory()
-        engine_dir = os.path.join(working_dir, "engine_dir")
-        pathlib.Path(engine_dir).mkdir(parents=True, exist_ok=True)
-        if additional_context is None:
-            additional_context = self.execution_state.additional_context if self.execution_state is not None else None
-        elif self.execution_state is not None and self.execution_state.additional_context is not None:
-            additional_context = {**self.execution_state.additional_context, **additional_context}
-        exec_state = ExecutionState(
-            mode=mode, working_dir=working_dir, engine_dir=engine_dir, additional_context=additional_context
-        )
+    def with_file_access(self, fa: _data_proxy.FileAccessProvider) -> Builder:
+        return self.new_builder().with_file_access(fa)
 
-        # If a wf_params object was not given, use the default (defined at the bottom of this file)
-        new_ctx = FlyteContext(
-            parent=self,
-            execution_state=exec_state,
-            user_space_params=execution_params or default_user_space_params,
-        )
-        FlyteContext.OBJS.append(new_ctx)
-        try:
-            yield new_ctx
-        finally:
-            FlyteContext.OBJS.pop()
+    def with_serialization_settings(self, ss: SerializationSettings) -> Builder:
+        return self.new_builder().with_serialization_settings(ss)
 
-    @property
-    def compilation_state(self) -> Optional[CompilationState]:
-        if self._compilation_state is not None:
-            return self._compilation_state
-        elif self._parent is not None:
-            return self._parent.compilation_state
-        else:
-            return None
+    def new_compilation_state(self, prefix: str = "") -> CompilationState:
```

Comment:
I know i'm mostly to blame for this, but with this change, there's now
* `FlyteContext.with_compilation_state`
* `FlyteContext.with_new_compilation_state`
* `FlyteContext.new_compilation_state`
which are repeated for the Builder object
* `FlyteContext.Builder.with_compilation_state`
* `FlyteContext.Builder.with_new_compilation_state`
* `FlyteContext.Builder.new_compilation_state`

Can we at least get rid of `new_compilation_state`? It doesn't seem to do anything except call the constructor in both places.

Reply:
We can, I was just so afraid to get rid of it